pkg/common/ipsec/utils.go

```
@@ -17,13 +20,23 @@ const (
 
 func CountUniqueIPsecKeys(states []netlink.XfrmState) int {
 	keys := make(map[string]bool)
+	invalidStateFound := false
 	for _, s := range states {
-		if s.Aead == nil {
+		if s.Aead != nil && s.Auth == nil && s.Crypt == nil {
+			keys[string(s.Aead.Key)] = true
+			continue
+		}
+		if s.Aead == nil && s.Auth != nil && s.Crypt != nil {
+			// we want to count the number of unique (Auth, Crypt) tuples
+			key := fmt.Sprintf("%s:%s", string(s.Auth.Key), string(s.Crypt.Key))
+			keys[key] = true
 			continue
 		}
-		keys[string(s.Aead.Key)] = true
+		invalidStateFound = true
+	}
+	if invalidStateFound {
+		fmt.Fprintf(os.Stderr, "an unsupported XfrmStateAlgo combination has been found\n")
```

Since we know that this is an invalid state, why not also return an error in this case?

---

That was the initial idea but it causes regression: https://github.com/cilium/cilium/pull/29182#pullrequestreview-1741248733

---

Is there a way for us to mark XFRM states as being owned by Cilium?

---

I have no idea... (((

---

Not really, no, and it feels fairly out-of-scope for this PR anyway.

---

Error handling here would help to prevent related bugs in the future.

---

Feel free to send a PR or CFP with a proposed approach. The goal of this PR is to add support for non-AEAD keys, not to deal with concurrent software using XFRM.

---

External factors are one reason to add error handling, another is giving the agent the ability to detect and recover in such instances (including unintentional corruption by the agent itself).

With regards to the above introduced test failure, I wouldn't suggest that we change the existing logic to exit on an error condition, rather add an error return value and ignore it in all cases where it is currently ignored. This gives, at least, future consumers of this function the opportunity to implement error handling.

---

I would agree with @asauber here. This is an exported util function and the call site won't be able to assert the validity of the returned count should one or mode invalid entries are found in the slice. 
Not to mention, we are assuming we are on the console which seems odd from a convenience call perspective??

---

@asauber @derailed I'm going to add error to the function signature.
Are you okay if the function returns the IPsec key count even if an error was detected? or it should return 0?
Thanks!

---

Non-zero for other parameters is definitely fine

---

Done.